### ISSUE TITLE
Possible bug causing AttributeError

### ISSUE THREAD

**Issue by semyonsh** [sopel-irc/sopel#2217]
### Description
Exceptions now and then, appear seemingly randomly but obviously triggered by some event, maybe users parting/joining/setting away. 

### Reproduction steps
1. No exotic configuration, the bot is joined to two channels, of which one is quite active. 
2. I run Sopel directly from the repo folder, ```python3.8 sopel.py ```
3. In regards to point 2, it's noted that there has to have taken place " python setup.py build magic before at some point." in order for it to work 😅

I have some custom plugins that all run on a hook. 

### Expected behavior
No exceptions as the bot is operating no different (By me) as it did before cloning it recently again. 

### Logs
Sopel console example:
```
[2021-12-01 11:09:31,226] sopel.irc.backends   INFO     - Connection error...  
[2021-12-01 11:09:31,226] sopel.irc            ERROR    - Fatal error in core, please review exceptions log.
```
Exception log: 
```
Last Line:
@time=2021-11-30T23:41:32.001Z;account=Anon :anon!anon@user/anon QUIT :Ping timeout: 120 seconds

[2021-12-01 00:41:32,017] sopel.exceptions     ERROR    - Fatal error traceback
Traceback (most recent call last):
  File "/usr/lib/python3.8/asyncore.py", line 83, in read
    obj.handle_read_event()
  File "/usr/lib/python3.8/asyncore.py", line 420, in handle_read_event
    self.handle_read()
  File "/usr/lib/python3.8/asynchat.py", line 171, in handle_read
    self.found_terminator()
  File "/home/semyon/sopel/sopel/irc/backends.py", line 229, in found_terminator
    self.bot.on_message(line)
  File "/home/semyon/sopel/sopel/irc/__init__.py", line 267, in on_message
    self.dispatch(pretrigger)
  File "/home/semyon/sopel/sopel/bot.py", line 802, in dispatch
    if "time" in pretrigger.tags and pretrigger.sender in self.channels:
  File "/home/semyon/sopel/sopel/tools/__init__.py", line 615, in __contains__
    return super().__contains__(Identifier(key))
  File "/home/semyon/sopel/sopel/tools/__init__.py", line 276, in __new__
    s._lowered = Identifier._lower(identifier)
  File "/home/semyon/sopel/sopel/tools/__init__.py", line 300, in _lower
    low = identifier.lower().replace('[', '{').replace(']', '}')
AttributeError: 'NoneType' object has no attribute 'lower'
[2021-12-01 00:41:32,018] sopel.exceptions     ERROR    - ----------------------------------------
[2021-12-01 00:42:24,305] sopel.exceptions     ERROR    - Fatal error in core, handle_error() was called.
Buffer:
```

### Environment
- Sopel `.version`: best guess: a8dcb76 
- Sopel installed via: source, git clone
- Python version: 3.8.10
- Operating system: Ubuntu 20.04.3 LTS
- IRCd `/version`:  InspIRCd-3
- Relevant plugins:  Only reload and admin are loaded from Sopel's native plugins. I run about 10 custom plugins that are all called by a hook.

### Notes
Exirel sent me from IRC 😄 If any more info is needed or logging I'll be happy to supply them, but in the meantime I will start using the pip version instead of running from source.

**Comment by semyonsh** [comment]
Pulled and now on commit 03b5cae, same exception.

```
Last Line:
@time=2021-12-01T16:36:55.658Z;account=Anon :anon!anon@snoonet.org/user/anon AWAY

[2021-12-01 17:36:55,668] sopel.exceptions     ERROR    - Fatal error traceback
Traceback (most recent call last):
  File "/usr/lib/python3.8/asyncore.py", line 83, in read
    obj.handle_read_event()
  File "/usr/lib/python3.8/asyncore.py", line 420, in handle_read_event
    self.handle_read()
  File "/usr/lib/python3.8/asynchat.py", line 171, in handle_read
    self.found_terminator()
  File "/home/semyon/sopel/sopel/irc/backends.py", line 229, in found_terminator
    self.bot.on_message(line)
  File "/home/semyon/sopel/sopel/irc/__init__.py", line 267, in on_message
    self.dispatch(pretrigger)
  File "/home/semyon/sopel/sopel/bot.py", line 802, in dispatch
    if "time" in pretrigger.tags and pretrigger.sender in self.channels:
  File "/home/semyon/sopel/sopel/tools/__init__.py", line 615, in __contains__
    return super().__contains__(Identifier(key))
  File "/home/semyon/sopel/sopel/tools/__init__.py", line 276, in __new__
    s._lowered = Identifier._lower(identifier)
  File "/home/semyon/sopel/sopel/tools/__init__.py", line 300, in _lower
    low = identifier.lower().replace('[', '{').replace(']', '}')
AttributeError: 'NoneType' object has no attribute 'lower'
[2021-12-01 17:36:55,668] sopel.exceptions     ERROR    - ----------------------------------------
```

**Comment by half-duplex** [comment]
Repro:
- Join Sopel to a channel you're in
- Do `/away bye` to mark yourself away
  - pretrigger.sender for this message is `"bye"` which is nonsense
- Do `/away` to mark yourself un-away
  - pretrigger.sender for this message is `None`, which causes an exception

**Comment by Exirel** [comment]
I can't reproduce that anymore, so this might have been fixed with latest from source. If someone has more information about that error, I'll look into it.

**Comment by dgw** [comment]
Based on spending a few minutes throwing together the below test cases (in `test/test_trigger.py`), I rather doubt that this is magically fixed on master.

```py
def test_away_pretrigger_away(nick):
    line = ':Foo!foo@example.com AWAY :gone fishing'
    pretrigger = PreTrigger(nick, line)
    assert pretrigger.tags == {}
    assert pretrigger.hostmask == 'Foo!foo@example.com'
    assert pretrigger.line == line
    assert pretrigger.args == ['gone fishing']
    assert pretrigger.text == 'gone fishing'
    assert pretrigger.plain == 'gone fishing'
    assert pretrigger.event == 'AWAY'
    assert pretrigger.nick == Identifier('Foo')
    assert pretrigger.user == 'foo'
    assert pretrigger.host == 'example.com'
    assert pretrigger.sender is None


def test_away_pretrigger_back(nick):
    line = ':Foo!foo@example.com AWAY'
    pretrigger = PreTrigger(nick, line)
    assert pretrigger.tags == {}
    assert pretrigger.hostmask == 'Foo!foo@example.com'
    assert pretrigger.line == line
    assert pretrigger.args == []
    assert pretrigger.text == ''
    assert pretrigger.plain == ''
    assert pretrigger.event == 'AWAY'
    assert pretrigger.nick == Identifier('Foo')
    assert pretrigger.user == 'foo'
    assert pretrigger.host == 'example.com'
    assert pretrigger.sender is None
```

Running against 71380df85212747a6c02d8c37c9ffb7836001278, for an AWAY marking the user as away (with a message), the sender is still set to the message. Without a message, some more weird stuff happens (e.g. `trigger.text == 'AWAY'`, somehow).

I would expect `trigger.sender` to be one of two values, depending on the convention we want:

- `trigger.sender` for non-channel-specific commands like AWAY should be `None`
- `trigger.sender` for non-channel-specific commands like AWAY should be the same as `trigger.nick`

And underneath all of this, Sopel shouldn't receive any AWAY commands like the ones shown above in OP's logs unless it's on an IRC server with the `away-notify` CAP turned on. That's something to watch out for if trying to test this bug live.

**Comment by Exirel** [comment]
That's a good question, however, the bug is more about `if "time" in pretrigger.tags and pretrigger.sender in self.channels:` raising an exception, because it tries to check if `None` exists in a `SopelMemory`.

That bug, in particular, has been fixed by 7c2bb07803067eb97fe54be0ee9d9b06cf57cf5e, so yes, the error disappeared on master.

As for the AWAY having or not having a `trigger.sender`... I think that's for another issue.

**Comment by dgw** [comment]
Right, then. Let's close this and continue talking about sender wonkiness in #2331.
